XWIKI-22958: The page picker is very slow on large wikis

michitux opened the pull request:
# Jira URL

<!-- Add the link to the corresponding JIRA issue referenced in a commit message. Unless this is a [Misc] commit,
see https://dev.xwiki.org/xwiki/bin/view/Community/DevelopmentPractices#HRule:Don27tcreateunnecessaryissues
-->

https://jira.xwiki.org/browse/XWIKI-22958

# Changes

## Description

<!-- Describe the main changes brought in this PR. -->

* First draft of a search implementation using the Solr index.
* Adjust tests to wait for Solr indexing.
* Update test to expect that searching by space returns results again.

## Clarifications

<!-- Provide extra hints to make it easier to understand the PR. Those could be:
* Explanation of choices made in this PR
* Anchor towards extra resources needed to understand the context of this PR (e.g., a forum proposal).
* Links to other issues this issue depends on
-->

* I haven't fully understood yet how localization works in Solr, I think I would need some pointers how to filter correctly by locale.
* The behavior has been intentionally changed again as https://jira.xwiki.org/browse/XWIKI-16934 and the last comment on https://jira.xwiki.org/browse/XWIKI-20632 indicated that it might actually be desired to search in the space, in particular if we use Solr.
* I haven't implemented support for sorting yet, and I'm not sure if we should: sorting by the quality of the match seems much better than sorting by any property and should provide much better suggestions in the page picker. If we want to have sorting, we would at least need to use "AND" as default query operator I fear.
* I haven't added a wildcard `*` for the content as I fear this would have a bigger impact on performance, but I have no data to back this, we could also add one.
* I haven't added a wildcard before the search string for performance reasons.
* The old database search implementation still exists as I'm wondering if we should make it configurable where the search should happen for backwards compatibility/to offer an easy way back in case somebody should miss an important feature.
* Object and space search haven't been addressed yet as they seem more difficult to migrate.

# Screenshots & Video

<!-- If this PR introduces any UI change, it's recommended to highlight it with before/after screenshots 
or even a screen recording for complex interactions. 
-->

No direct UI change.

# Executed Tests

<!-- Especially important for regression fixes. 
Indicate how changes were tested (e.g., what maven commands were run to validate them).
-->

```
mvn clean install -Pquality,integration-tests,docker -pl :xwiki-platform-rest-server,:xwiki-platform-rest-test-tests
```

# Expected merging strategy

* Prefers squash: Yes <!-- No — Explain why. -->
* Backport on branches:
  * Seems a bit dangerous but might still be important, not sure.

michitux commented on xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/internal/resources/BaseSearchResult.java:
```diff
@@ -137,6 +155,63 @@ protected List<SearchResult> search(List<SearchScope> searchScopes, String keywo
+
+        List<String> filterQueries = new ArrayList<>();
+        filterQueries.add("type:DOCUMENT");
+        filterQueries.add("hidden:false");
```
I guess this should depend on the user preference. We don't have a generic hidden document query filter for Solr queries, do we?

mflorea replied:
We don't. It would have to extend ``AbstractHiddenFilter``.

mflorea commented on xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/internal/resources/BaseSearchResult.java:
```diff
@@ -137,6 +155,63 @@ protected List<SearchResult> search(List<SearchScope> searchScopes, String keywo
+        }
+
+        Query query = this.queryManager.createQuery(queryString, "solr");
+        query.setLimit(number * 2);
```
This deserves some explanation.

michitux replied:
I copied this from the old database code where the explanation was that we could match both the original and the translation. It's entirely possibly that we actually don't need this here because we match the documents correctly.

mflorea commented on xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/internal/resources/BaseSearchResult.java:
```diff
@@ -137,6 +155,63 @@ protected List<SearchResult> search(List<SearchScope> searchScopes, String keywo
+        if (Boolean.TRUE.equals(isLocaleAware)) {
+            // TODO: this might not be the right way to filter by language.
+            Locale currentLocale = this.localizationContext.getCurrentLocale();
+            filterQueries.add("locale:(\"\" OR %s OR %s)".formatted(currentLocale, currentLocale.getLanguage()));
```
That's not what we do in https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-ui/src/main/resources/XWiki/SuggestSolrMacros.xml#L108 . If you want to match the parent locale as well then you need to do as indicated on https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-ui/src/main/resources/Main/SolrSearchMacros.xml#L746  i.e. use the ``locales`` (plural). But since this is not enabled for the default search page, I think it's enough to do:

```suggestion
            filterQueries.add("locale:(\"\" OR %s)".formatted(currentLocale));
```

michitux commented:
Updates:

* I've implemented hidden filtering based on the user's preferences.
* I've implemented sorting for those fields that are available for sorting in Solr.
* I've changed the locale support to use the `locales` field as this seems to match much better what we had before.
* I've removed the database implementation. Having two implementations that we need to maintain while only one is actually used doesn't seem good to me.
* I've improved wildcard search to get the last token using the same standard tokenizer as Solr. For titles, I've changed the field for wildcard search to the field that is index as general text to avoid having a non-standard tokenizer.
* I've added a special case to match the document reference as a whole, this is to support the use case when the user input is a full document reference. This doesn't fully work in the page picker as it might not actually show the result as the page picker performs filtering, too. I don't know if this is a good idea, but it seemed like an interesting feature for the page picker as some people use it with full document reference.

Still open:

* [ ] Implement a special case for empty search queries to return the last modified documents when no order is given, or simply return the first documents according to the specified order.

michitux commented:
I've refactored the code to make the backend that is used for the wiki and space search endpoint configurable to allow admins to easily switch back to the previous behavior in case they're relying on it. Further, there is less code duplication now.

These changes are in the last commit (I've force-pushed to rebase on master, but I haven't modified any existing commits):

* Introduce a new (internal) role for keyword search with two
  implementations (database and solr) that can be changed based on
  a new configuration option to allow switching back to the database
  implementation.
* Introduce a parameter object to avoid having methods with a huge
  number of parameters.
* Move the Solr-based keyword search to the
  `xwiki-platform-search-solr-rest` module.
* Make the existing code for transforming Solr results into search
  result objects re-usable and re-use it.
* Clean up the code to reduce Checkstyle violations, remove the global
  Checkstyle exclude for BaseSearchResult.
